Read mff evoked

ephathaway opened the pull request:
This PR is the pre-PR so that we can review the code amongst BEL engineers before submitting the actual PR to mne-tools.

In this PR, we begin to address #8038 by implementing reading of averaged MFF files. We add a function `mne.io.read_evokeds_mff` which takes in an averaged MFF file and returns an `EvokedArray` object with the data for each condition specified. Reading of signal data and some metadata is done through [mffpy](https://github.com/BEL-Public/mffpy), which is made a dependency of mne.

There are a few things we will need to sort out before this is ready to merge that I will list right off the bat.
1. mne testing datasets are stored in a separate [repo](https://github.com/mne-tools/mne-testing-data). I am not sure what their process is for adding new test datasets, so for now I include my test averaged MFF file in this repository.
2. Right now there is no check to make sure the input MFF file is averaged. This will be important for preventing users from trying to read a segmented file with the `mne.io.read_evokeds_mff` function. I suggest we use the information in `history.xml`, which contains info on all of the tools run on the MFF file. We currently don't have a way to parse `history.xml`, so I suggest we first implement this capability in mffpy and then use it to read the info in mne.
3. The number of segments comprising each average is specified in `categories.xml` in an averaged MFF file, but again, we do not currently have the ability to parse this info. I suggest we implement this capability in mffpy as well.

ephathaway committed:
ENH: support reading of averaged MFF files

Here we add a function `mne.io.read_evokeds_mff`
to read averaged MFF files. This function works
similarly to `mne.evoked.read_evokeds` by returning
the averaged data and metadata from one or more
specified `condition`(s).

We make use of the mffpy[https://github.com/BEL-Public/mffpy]
package for reading the signal data and some of
the metadata in the MFF file.

ephathaway committed:
Add test for `mne.io.read_evokeds_mff`

To test reading of averaged MFF files,
we add a new test MFF file
`example_mffs/test_egi_evoked.mff` and
two .txt files with the signals to compare
against.

These eventually need to be moved to the test
data repo, which I believe would be
https://github.com/mne-tools/mne-testing-data
in this case.

ianbrown9475 commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,175 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+        condition = list(cats.keys())
+    elif not isinstance(condition, list):
+        condition = [condition]
+    logger.info('Reading %s evoked datasets from %s ...'
```
I think the first `%s` should be a `%d`

ephathaway replied:
Good catch.

ianbrown9475 commented on mne/io/egi/egimff.py:
```diff
@@ -9,15 +9,18 @@
 
 import numpy as np
 
+from mffpy import Reader
```
The project seems to follow PEP 8 so I think this line should be directly under `import numpy as np` https://www.python.org/dev/peps/pep-0008/#imports

ephathaway replied:
In MNE's instructions for contributing they say to order imports like this:

Python built-in (os, copy, functools, etc)
standard scientific (numpy as np, scipy.signal, etc)
others
MNE-Python imports (e.g., from .pick import pick_types)

I'm thinking `mffpy` falls under "others" in this case.

ianbrown9475 replied:
Ah, didn't see that. Good catch

ianbrown9475 commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,175 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+    evoked : EvokedArray or list of EvokedArray
+        The evoked dataset(s); one EvokedArray if condition is int or str,
+        or list of EvokedArray if condition is None or list.
+
```
Could you also add a `Raises` section?

ephathaway replied:
Yep, good call.

ianbrown9475 commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,175 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+                range(mff.num_channels['EEG'])]
+    ch_names.extend(mff_info['pns_names'])
+    info = create_info(ch_names, mff.sampling_rates['EEG'], ch_types)
+    info['nchan'] = sum(n for n in mff.num_channels.values())
```
Do you need the list comprehension here or can you just do `sum(mff.num_channels.values())`?

ephathaway replied:
Nice, yeah it doesn't need the list comprehension.

ianbrown9475 commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,175 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+    # Get calibration info
+    gains = _get_gains(op.join(fname, mff_info['info_fname']))
+    if mff_info['value_range'] != 0 and mff_info['bits'] != 0:
+        cals = [mff_info['value_range'] / 2 ** mff_info['bits'] for i
+                in range(len(mff_info['chan_type']))]
```
Since the expression in this list comprehension isn't dependent on the list being constructed from `for i in range(len(mff_info['chan_type']))`, I think you can replace this with
```python3
cals = [mff_info['value_range'] / 2 ** mff_info['bits']] * len(mff_info['chan_type'])
```

This should be more performant too as you avoid having to construct a separate list.

ephathaway replied:
Yes! I knew there was an easier way to write that line. Thanks

ephathaway committed:
Set mffpy requirement to latest release

ianbrown9475 reviewed (changes requested):
Nice work, I ran the tests and it looks like the ones you added/modified are passing but there are others that aren't. I think some of them were already failing in the upstream code though. Could you look at fixing the CI action that's failing and fixing the linter errors? Looks like there's some more information on all of this [here](https://mne.tools/dev/install/contributing.html#contributing).

ephathaway committed:
STY: docstring and code style

ephathaway commented:
CIs appear to be passing now. I ran flake8 and docstyle tests and those look good. Some tests failing on my end as well, but they all look unrelated to my additions. Looks like they have to do with missing test files, so I might just be missing some tests datasets that I was supposed to install.

jusjusjus commented on requirements.txt:
```diff
@@ -28,3 +28,5 @@ imageio-ffmpeg>=0.4.1
 pyvistaqt>=0.2.0
 tqdm
+mffpy==0.5.3
+cached_property_bel
```
cached_property_bel>=1.0.1

jusjusjus commented on mne/io/egi/tests/test_egi.py:
```diff
@@ -288,4 +292,35 @@ def test_io_egi_crop_no_preload():
+    (1, 'Category 2', egi_txt_evoked_cat2_fname,
+     ['E257', 'EMG'])
+])
+def test_io_egi_evoked_mff(idx, cond, signals, bads):
```
`evokeds`

ephathaway replied:
Good catch.

jusjusjus commented on mne/io/egi/tests/test_egi.py:
```diff
@@ -29,6 +29,10 @@
 egi_pause_w1337_fname = op.join(egi_path, 'w1337_20191014_105416.mff')
+# Example MFFs are stored in a separate repository mne-tools/mne-testing-data
+egi_mff_evoked_fname = op.join('example_mffs', 'test_egi_evoked.mff')
+egi_txt_evoked_cat1_fname = op.join('example_mffs', 'test_egi_evoked_cat1.txt')
+egi_txt_evoked_cat2_fname = op.join('example_mffs', 'test_egi_evoked_cat2.txt')
```
Should we store the test data as npz to reduce the memory footprint, or is that not how they recommend to do it?

ephathaway replied:
It looks like their other signal data to compare against is in .txt format. I'm sure they will give us some instruction on how they want the test data formatted when we submit the PR.

jusjusjus commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,182 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+    AssertionError
+        If `fname` has file extension other than '.mff'.
+    AssertionError
+        If no categories.xml file in `fname` directory.
```
Should be `ValueError`.

jusjusjus commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,182 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+@verbose
+def read_evokeds_mff(fname, condition=None, channel_naming='E%d',
+                     baseline=None, verbose=None):
+    """Read averaged MFF file as evoked(s).
```
as list of `EvokedArray`.

jusjusjus commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,182 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+    mff = Reader(fname)
+    cats = mff.categories.categories
+
+    if type(condition) == str:
```
Best to use `isinstance`.  I checked what they use:

```bash
$ grep 'if isinstance(' mne/**/* | wc
391 occurances
$ grep 'if type(' mne/**/* | wc
3 occurances
```

ephathaway replied:
Yeah, good call.

jusjusjus commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,182 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+        # Condition is interpreted as category name
+        target_cat = condition
+        target_epoch = mff.epochs[target_cat]
+    elif type(condition) == int:
```
ibid

jusjusjus commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,182 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+    if 'gcal' in gains:
+        cals *= gains['gcal']
+    if 'ical' in gains:
+        pass  # XXX: currently not used
```
Either drop it, throw a warning once (https://docs.python.org/3/library/warnings.html), or throw a `NotImplementedError`

jusjusjus commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,182 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+    for i_ch, ch_name in enumerate(mff_info['pns_names']):
+        idx = ch_names.index(ch_name)
+        ch_type = mff_info['pns_types'][i_ch]
+        ch_kind = FIFF.FIFFV_BIO_CH
+        if ch_type == 'ecg':
+            ch_kind = FIFF.FIFFV_ECG_CH
+        elif ch_type == 'emg':
+            ch_kind = FIFF.FIFFV_EMG_CH
```
```python
type_to_kind_map = {
   'ecg': FIFF.FIFFV_ECG_CH
...
}
type_to_kind_map.get(ch_type, FIFF.FIFFV_BIO_CH)
```
Also you can do
`from FIFF import FIFFV_BIO_CH,...`
Their naming is a bit redundant here.

ephathaway replied:
Nice. That `dict.get()` function is really handy.

jusjusjus commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,182 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+
+    if type(condition) == str:
+        # Condition is interpreted as category name
+        target_cat = condition
```
`category`

jusjusjus commented on mne/io/egi/egimff.py:
```diff
@@ -686,3 +684,182 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
+    if type(condition) == str:
+        # Condition is interpreted as category name
+        target_cat = condition
+        target_epoch = mff.epochs[target_cat]
```
`epoch`

jusjusjus reviewed (changes requested):
Hi Evan,  well done.  You definitely hit the mark wrt their code base, but I would like this commit to be a little different here an there.

* when including mffpy as a dependency, it would have been awesome if cached-property-bel automatically is added as a sub-dependency.  Let's hear what they say about this, but get ready to include that into the pip wheel.  We should also edit the "server_environment.yml"

* The test data you added is good, and it's great that you check that mffpy actually reads it.  I think we might need to clean up some of the viewer configuration files though, which are unnecessary for mffpy, e.g. "lastSettings.xml", "recordingSettings.xml".  I don't think we support reading these.

* You rightly named the child `read_evokeds_...`, b/c it returns mutliple.  Instead of messing around with _a list or an instance_ I would just return a list with one element.  The branch (if else) will cause downstream branches by the guy that uses the API, otherwise.

* It's good that you check the function input which immediately enhances the usability of the API.  If a user inputs a wrong file than it'll come around and tell him.  It's good to make these error messages very precise; tell the user what he did wrong.  Also, throw a `ValueError` instead of an `AssertionError` (https://docs.python.org/3/library/exceptions.html#ValueError).

* That's a matter of taste, but in general i'm not a big fan of abbreviations.  If you want to avoid long names, then, one might be able to use context.  One thing that would help in creating this context are functions.  I think one could split `_read_evoked_mff` into 3-4 sub-functions that a called in sequence.  In essence you do this, and you even write little documentation strings for each function, but by not actually creating a function your code always drags the whole context with it.

ephathaway commented:
> * You rightly named the child `read_evokeds_...`, b/c it returns mutliple.  Instead of messing around with _a list or an instance_ I would just return a list with one element.  The branch (if else) will cause downstream branches by the guy that uses the API, otherwise.

I decided to return a single `Evoked_Array` if just reading one dataset instead of a list to stay consistent with `mne.read_evokeds` (for reading evoked .fif files).

ephathaway committed:
STY: code consistency and efficiency

Here we make some stylistic changes such as
variable naming convention to make the code
more readable and consistent with the MNE style.

We implement two new private functions in
`mne/io/egi/egimff.py` to replace some code
that was redundant between class `RawMff`
and `_read_evoked_mff`.

ephathaway committed:
STY: new line at end of

jusjusjus reviewed (approved):
Thanks for addressing my comments.  Let's see what they say about this.

ianbrown9475 reviewed (approved):
Looks good to me!

ephathaway commented:
> * Right now there is no check to make sure the input MFF file is averaged. This will be important for preventing users from trying to read a segmented file with the `mne.io.read_evokeds_mff` function. I suggest we use the information in `history.xml`, which contains info on all of the tools run on the MFF file. We currently don't have a way to parse `history.xml`, so I suggest we first implement this capability in mffpy and then use it to read the info in mne.
> * The number of segments comprising each average is specified in `categories.xml` in an averaged MFF file, but again, we do not currently have the ability to parse this info. I suggest we implement this capability in mffpy as well.

@ianbrown9475 @jusjusjus thank you both for reviewing this! Your feedback was very helpful. Before I submit a PR to mne-tools, I would like to address the above issues. I will do some work on `mffpy` next week and hopefully have PR submitted by Friday.

jusjusjus commented:
Hey Evan, I think you would be right to check if the file is averaged beforehand.

ephathaway committed:
ENH: check if input file is averaged

jusjusjus reviewed (approved):
Looks good, great job.

ianbrown9475 reviewed (approved):
Nice, looks good

jusjusjus committed:
ENH: Update mffpy to >=0.5.5

jusjusjus committed:
ENH: import mffpy dynamically

jusjusjus committed:
ENH: Add version added string

jusjusjus committed:
FIX: Fix evoked test to align mne policy

ephathaway committed:
FIX: organize testing data

ephathaway committed:
ENH: apply reference info when reading MFF

ephathaway committed:
DOC: update changelog

ephathaway committed:
FIX: loose version for mffpy requirement

ephathaway committed:
DOC: fix doc string style

ephathaway committed:
FIX: python ref for

ephathaway committed:
DOC: new contributor changelog entry

ephathaway committed:
DOC: make parameter references render as code

ephathaway committed:
Move test signals to mne-testing-data

ephathaway committed:
ENH: check channel types and names

ephathaway committed:
FIX: require mffpy 0.5.5 for test

ephathaway committed:
FIX: pull in version bug fix in mffpy

ephathaway committed:
ENH: read nave from categories.xml

ephathaway committed:
Improve test coverage

ephathaway committed:
DOC: fix version added

ephathaway committed:
STY: use f-strings and replace asserts

ephathaway committed:
ENH: test error messages for bad inputs

ephathaway committed:
FIX: return list if condition is list

jusjusjus reviewed (approved):
Looks good

ephathaway commented:
> Looks good

Thanks. I still think we should merge this into `mne-tools/mne-python` master branch when it's approved and then pull those changes into our master.

ianbrown9475 commented:
@ephathaway Yeah I think that's pretty common for these kind of situations. Are you going to open a PR in `mne-tools/mne-python`?

ephathaway commented:
> @ephathaway Yeah I think that's pretty common for these kind of situations. Are you going to open a PR in `mne-tools/mne-python`?

It's been open for a while now. Gone through several rounds of review, but I think it's close. https://github.com/mne-tools/mne-python/pull/8354

jusjusjus commented:
> > Looks good
> 
> Thanks. I still think we should merge this into `mne-tools/mne-python` master branch when it's approved and then pull those changes into our master.

Sure, I think we discussed in detail how we should have done it.

- fork upstream/master -> (our) origin/master
- develop in a branch origin/feature
- PR origin/feature -> origin/master for internal review
- rebase origin/master on top of upstream/master 
- PR origin/master -> upstream/master
- rebase as needed until merged

It's at your discretion how to continue on the current path though.  I'm happy to discuss options.

ephathaway commented:
> It's at your discretion how to continue on the current path though. I'm happy to discuss options.

Yeah, I think it makes sense to continue on the current path this time around since it's almost ready to merge. In my opinion, it's nice to submit the PR with origin/feature_branch -> upstream/master because then it's easy to push requested changes directly to origin/feature_branch instead of heaving to create a new branch every time we want to make changes if we were doing origin/master -> upstream/master.

ephathaway committed:
DOC: reference mffpy dependency in README
